### PR TITLE
[WIP] Implement opt-in graceful delete extension for custom resources

### LINKED ISSUE #56567: Grace period on DeleteOptions has no effect on CRs
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

We use a CR with a custom finalizer to get support for graceful deletes. While the crd apiserver seems to consider the deleteGracePeriod to some degree, it seems that it is not updating `metadata.deletionGracePeriodSeconds`. Therefore our finalizer has no way to differentiate between a forceful delete request `--grace-period 0` or a graceful delete request.

A follow up issue seems to be that `--grace-period` is not sent by kubectl if it talks to a cr. Therefore I tried it with a DeleteOptions object on a plain curl call too.

**What you expected to happen**:

I would expect that I can implement graceful delete based on a finalizer on my CR, in combination with sending `gracePeriodSecons` to the apiserver. This allows me to distinguish between a force delete and a graceful delete.

**How to reproduce it (as minimally and precisely as possible)**:

Here the kubectl log:

```bash
$ delete vms  testvm --grace-period 20 -v9
Request Body: {"orphanDependents":false}
I1129 12:13:33.298274     728 round_trippers.go:417] curl -k -v -XDELETE  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: .kubectl/v1.8.4 (linux/amd64) kubernetes/9befc2b" https://192.168.121.164:6443/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm
I1129 12:13:33.312421     728 round_trippers.go:436] DELETE https://192.168.121.164:6443/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm 202 Accepted in 14 milliseconds
I1129 12:13:33.312486     728 round_trippers.go:442] Response Headers:
I1129 12:13:33.312542     728 round_trippers.go:445]     Content-Type: application/json
I1129 12:13:33.312549     728 round_trippers.go:445]     Content-Length: 1198
I1129 12:13:33.312554     728 round_trippers.go:445]     Date: Wed, 29 Nov 2017 11:13:33 GMT
I1129 12:13:33.312621     728 request.go:836] Response Body: {"apiVersion":"kubevirt.io/v1alpha1","kind":"VirtualMachine","metadata":{"clusterName":"","creationTimestamp":"2017-11-29T11:07:54Z","deletionGracePeriodSeconds":0, [...]}
```
As you can see, no `gracePerodSeconds` is sent to the apiserver and the resulting delete response has `deletionGracePeriodSecons` set to 0.

And here the plain curl call:

```bash
$ curl -k -v -XDELETE  -H "Content-Type: application/json" -H "Accept: application/json" -H "User-Agent: .kubectl/v1.8.4 (linux/amd64) kubernetes/9befc2b" http://192.168.200.2:8184/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm --data '{"orphanDependents":false, "gracePeriodSeconds": 20}'
{"metadata":{"name":"testvm","namespace":"default","selfLink":"/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm","uid":"b7e81365-d4f6-11e7-8ec0-525400ea43c6","resourceVersion":"54319","creationTimestamp":"2017-11-29T11:16:17Z","deletionTimestamp":"2017-11-29T11:16:24Z","deletionGracePeriodSeconds":0, [...]}
```
This time I made sure by hand, that `gracePeriodSecons` is sent to the apiserver, but the resulting object still show 0 for `deletionGracePeriodSeconds`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

It extends CRDs with graceful delete support. This allows similar graceful delete capabilities for extensions like they exist for pods. An example may be a VirtualMachine workload type implemented via CRs, where a user would simply expect that `kubectl delete vms myvm --grace-period 0` or `kubectl vms myvm --grace-period 0 --force` works.

A user can opt-in to graceful-delete support by registering a crd  like this:

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: foos.samplecontroller.k8s.io
spec:
  group: samplecontroller.k8s.io
  version: v1alpha1
  names:
    kind: Foo
    plural: foos
  scope: Namespaced
  deleteStrategy:
    defaultTerminatingGracePeriodSeconds: 30
    terminatingGracePeriodSecondsPath: ".spec.terminatingGracePeriodSeconds"
```

If no `spec.deleteStrategy` is specified, then the usual force delete will happen during deletes.

The implementation tries to follow the same pattern like @nikhita used to implement subresources.

**Which issue(s) this PR fixes**:
Fixes #56567

**Special notes for your reviewer**:

POC for the second option mentioned in https://github.com/kubernetes/kubernetes/issues/56567#issuecomment-384232263/.

 * It would also be possible to add optional support for immediate-deletes based on an values in a configurable json path (e.g. .status.phase = [failed|succeeded]), like it exists for pods.
 * Most storage tests pass, but a CR graceful delete differs to a degree from pod graceful delete behaviour, so distinguishing between CR-graceful-delete tests and pod-graceful-delete tests will be necessary. Commented the currend pod-specific tests out, to illustrate what is currently failing.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by sttts** [comment]
/cc @kubernetes/sig-api-machinery-api-reviews 

@deads2k @lavalamp @liggitt @caesarxuchao please take a look at this proposal. Has there been any discussion about finalizers vs. graceful deletion anywhere? The later predates the former by quite some time. Finalizers could simulate graceful deletion.

**Comment by lavalamp** [comment]
/assign @yliaog @caesarxuchao 

Let's make sure that question gets answered before we merge this :)

**Comment by caesarxuchao** [comment]
I commented in the issue: https://github.com/kubernetes/kubernetes/issues/56567#issuecomment-384783547

**Comment by lavalamp** [comment]
/hold

We talked about this on the SIG. Let's have some more discussion (in the issue) before we commit ourselves to supporting this concept for every resource for all of eternity.

**Comment by rmohr** [comment]
/cc @sttts 

Updated everything. I would want to give this another try. @sttts I however have issues to understand under which circumstances the rest API should return a Status with success and when the object. I could not really make out the pattern when comparing the deletion of custom resources and pods.

File: staging/src/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go [hunk 1]
```diff
@@ -835,7 +835,7 @@ func (t *Tester) testDeleteNoGraceful(obj runtime.Object, createFn CreateFunc, g
 	}
 	defer t.delete(ctx, foo)
 	objectMeta := t.getObjectMetaOrFail(foo)
-	opts := metav1.NewDeleteOptions(10)
+	opts := metav1.NewDeleteOptions(0)
```
Comment (rmohr) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go]: It looked strange that the dryrun call should delete non-graceful although a grace period is set in the options ... What should be the correct behaviour for Pods and CRs?

**Pull request closed without merging**